var/spack/repos/builtin/packages/python/package.py

```diff
@@ -303,6 +305,60 @@ def determine_variants(cls, exes, version_str):
             except ProcessError:
                 variants += '~tix'
 
+        # Get config args for shared, debug, and optimizations variant checks
+        sysconf_imp = 'import sysconfig'
+        if version < Version('2.7'):
+            sysconf_imp = 'import distutils.sysconfig as sysconfig'
+        conf_args = python(
+            '-c',
+            '%s; print(sysconfig.get_config_var("CONFIG_ARGS"))' % sysconf_imp,
+            output=str, error=str)
+
+        # check for shared variant
+        # defualt is on, so we check for the string to turn it off
+        # +shared implies +pic
+        if '--disable-shared' in conf_args:
+            variants += '~shared'
+        else:
+            variants += '+shared'
+            variants += '+pic'
+
+        # check for debug
+        # default is off, so we check for the string to turn it on
+        variants += '+debug' if '--with-pydebug' in conf_args else '~debug'
+
+        # Use config args to determine whether optimizations were used
+        opt_var = '~optimizations'
+        # only check for versions that support optimization
+        if version >= Version('3.5.3') or (version >= Version('2.7.13') and
+                                           version < Version('2.8')):
+            if '--enable-optimizations' in conf_args:
+                opt_var = '+optimizations'
+        variants += opt_var
+
+        # +ucs4 only compatible with python@:3.2
+        if version >= Version('3.3'):
+            variants += '~ucs4'
+        else:
+            # maxunicode is 0xFFFF for UCS2 and 0x10FFFF for UCS4
+            maxunicode = python('-c', 'import sys; print(sys.maxunicode)',
+                                output=str, error=str)
+            variants += '~ucs4' if maxunicode == 0xFFFF else '+ucs4'
+
+        # +pythoncommand does not apply to python@2
```

Comment:
Personally, I don't think we should set this variant for Python 2 at all.

Reply:
We have to set the variant, we don't (yet) have a mechanism to have the concretizer omit a variant for certain versions. `ucs4` has a similar issue, where in that case we raise exceptions if we try to build with it at certain versions.

We can either set it to the default in the detection logic, or we can let the concretizer fill it in. The latter would be appealing except for the fact that it looks like we're going to need to swap the python that Spack is run with into an already concrete spec as part of our bootstrapping process (to know which binary to download for clingo/libstdc++). That means we need to have all the variants filled out by the detection logic.

Reply:
Hmm, in that case maybe just enable it for Python 2 since it's on by default and Python 2 installs a `python` executable.